[ICS Reply] Server fail issue when trying an Event Reply jmap method on prod

After latest release 0.10.1, when trying to use the new event reply feature on tmail in oncommit-tmail.linagora.com, it fails:

![Image](https://github.com/linagora/tmail-backend/assets/9005025/f926bce4-f744-4b43-a0bc-9fc971e2bb5c)

Investigating logs is showing this:

"timestamp":"2024-06-07T07:33:48.054Z","level":"ERROR","thread":"sdk-async-response-1-20097","mdc":{"method":"CalendarEvent/accept","protocol":"JMAP","username":"rcordier@linagora.com"},"logger":"org.apache.james.jmap.method.Method","message":"Server side error executing a JMAP method","context":"default","exception":"reactor.core.Exceptions$RetryExhaustedException: Retries exhausted: 0/0\n\tat reactor.core.Exceptions.retryExhausted(Exceptions.java:306)\n\tat reactor.util.retry.RetryBackoffSpec.lambda$static$0(RetryBackoffSpec.java:68)\n\tat reactor.util.retry.RetryBackoffSpec.lambda$null$4(RetryBackoffSpec.java:560)\n\tat reactor.core.publisher.FluxConcatMapNoPrefetch$FluxConcatMapNoPrefetchSubscriber.onNext(FluxConcatMapNoPrefetch.java:183)\n\tat reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.onNext(FluxContextWrite.java:107)\n\tat reactor.core.publisher.SinkManyEmitterProcessor.drain(SinkManyEmitterProcessor.java:476)\n\tat reactor.core.publisher.SinkManyEmitterProcessor$EmitterInner.drainParent(SinkManyEmitterProcessor.java:620)\n\tat reactor.core.publisher.FluxPublish$PubSubInner.request(FluxPublish.java:874)\n\tat reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.request(FluxContextWrite.java:136)\n\tat reactor.core.publisher.FluxConcatMapNoPrefetch$FluxConcatMapNoPrefetchSubscriber.request(FluxConcatMapNoPrefetch.java:336)\n\tat reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.request(FluxContextWrite.java:136)\n\tat reactor.core.publisher.Operators$DeferredSubscription.request(Operators.java:1743)\n\tat reactor.core.publisher.FluxRetryWhen$RetryWhenMainSubscriber.onError(FluxRetryWhen.java:195)\n\tat reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onError(FluxOnErrorResume.java:106)\n\tat reactor.core.publisher.Operators.error(Operators.java:198)\n\tat reactor.core.publisher.MonoError.subscribe(MonoError.java:53)\n\tat reactor.core.publisher.Mono.subscribe(Mono.java:4512)\n\tat reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onError(FluxOnErrorResume.java:103)\n\tat reactor.core.publisher.SerializedSubscriber.onError(SerializedSubscriber.java:124)\n\tat reactor.core.publisher.FluxRetryWhen$RetryWhenMainSubscriber.whenError(FluxRetryWhen.java:228)\n\tat reactor.core.publisher.FluxRetryWhen$RetryWhenOtherSubscriber.onError(FluxRetryWhen.java:278)\n\tat reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.onError(FluxContextWrite.java:121)\n\tat reactor.core.publisher.FluxConcatMapNoPrefetch$FluxConcatMapNoPrefetchSubscriber.maybeOnError(FluxConcatMapNoPrefetch.java:326)\n\tat reactor.core.publisher.FluxConcatMapNoPrefetch$FluxConcatMapNoPrefetchSubscriber.innerError(FluxConcatMapNoPrefetch.java:297)\n\tat reactor.core.publisher.FluxConcatMap$ConcatMapInner.onError(FluxConcatMap.java:875)\n\tat reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.onError(FluxContextWrite.java:121)\n\tat reactor.core.publisher.MonoFlatMap$FlatMapMain.onError(MonoFlatMap.java:180)\n\tat reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.onError(MonoIgnoreThen.java:279)\n\tat reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.subscribeNext(MonoIgnoreThen.java:256)\n\tat reactor.core.publisher.MonoIgnoreThen.subscribe(MonoIgnoreThen.java:51)\n\tat reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:76)\n\tat reactor.core.publisher.FluxConcatMapNoPrefetch$FluxConcatMapNoPrefetchSubscriber.onNext(FluxConcatMapNoPrefetch.java:206)\n\tat reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.onNext(FluxContextWrite.java:107)\n\tat reactor.core.publisher.SinkManyEmitterProcessor.drain(SinkManyEmitterProcessor.java:476)\n\tat reactor.core.publisher.SinkManyEmitterProcessor$EmitterInner.drainParent(SinkManyEmitterProcessor.java:620)\n\tat reactor.core.publisher.FluxPublish$PubSubInner.request(FluxPublish.java:874)\n\tat reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.request(FluxContextWrite.java:136)\n\tat reactor.core.publisher.FluxConcatMapNoPrefetch$FluxConcatMapNoPrefetchSubscriber.request(FluxConcatMapNoPrefetch.java:336)\n\tat reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.request(FluxContextWrite.java:136)\n\tat reactor.core.publisher.Operators$DeferredSubscription.request(Operators.java:1743)\n\tat reactor.core.publisher.FluxRetryWhen$RetryWhenMainSubscriber.onError(FluxRetryWhen.java:195)\n\tat reactor.core.publisher.MonoFlatMap$FlatMapMain.secondError(MonoFlatMap.java:241)\n\tat reactor.core.publisher.MonoFlatMap$FlatMapInner.onError(MonoFlatMap.java:315)\n\tat reactor.core.publisher.MonoUsingWhen$MonoUsingWhenSubscriber.deferredError(MonoUsingWhen.java:278)\n\tat reactor.core.publisher.FluxUsingWhen$RollbackInner.onComplete(FluxUsingWhen.java:476)\n\tat reactor.core.publisher.MonoRunnable.subscribe(MonoRunnable.java:50)\n\tat reactor.core.publisher.FluxUsingWhen$UsingWhenSubscriber.onError(FluxUsingWhen.java:365)\n\tat reactor.core.publisher.MonoCompletionStage$MonoCompletionStageSubscription.apply(MonoCompletionStage.java:115)\n\tat reactor.core.publisher.MonoCompletionStage$MonoCompletionStageSubscription.apply(MonoCompletionStage.java:67)\n\tat java.base/java.util.concurrent.CompletableFuture.uniHandle(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source)\n\tat software.amazon.awssdk.utils.CompletableFutureUtils.lambda$forwardExceptionTo$0(CompletableFutureUtils.java:79)\n\tat java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source)\n\tat software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncApiCallMetricCollectionStage.lambda$execute$0(AsyncApiCallMetricCollectionStage.java:56)\n\tat java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source)\n\tat software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncApiCallTimeoutTrackingStage.lambda$execute$2(AsyncApiCallTimeoutTrackingStage.java:67)\n\tat java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source)\n\tat software.amazon.awssdk.utils.CompletableFutureUtils.lambda$forwardExceptionTo$0(CompletableFutureUtils.java:79)\n\tat java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source)\n\tat software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryingExecutor.maybeAttemptExecute(AsyncRetryableStage.java:103)\n\tat software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryingExecutor.maybeRetryExecute(AsyncRetryableStage.java:184)\n\tat software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryingExecutor.lambda$attemptExecute$1(AsyncRetryableStage.java:170)\n\tat java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.complete(Unknown Source)\n\tat software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage.lambda$execute$0(MakeAsyncHttpRequestStage.java:105)\n\tat java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.complete(Unknown Source)\n\tat software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage.completeResponseFuture(MakeAsyncHttpRequestStage.java:238)\n\tat software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage.lambda$executeHttpRequest$3(MakeAsyncHttpRequestStage.java:163)\n\tat java.base/java.util.concurrent.CompletableFuture.uniHandle(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture$Completion.run(Unknown Source)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)\n\tat java.base/java.lang.Thread.run(Unknown Source)\nCaused by: software.amazon.awssdk.services.s3.model.S3Exception: null (Service: S3, Status Code: 400, Request ID: null)\n\tat software.amazon.awssdk.protocols.xml.internal.unmarshall.AwsXmlPredicatedResponseHandler.handleErrorResponse(AwsXmlPredicatedResponseHandler.java:156)\n\tat software.amazon.awssdk.protocols.xml.internal.unmarshall.AwsXmlPredicatedResponseHandler.handleResponse(AwsXmlPredicatedResponseHandler.java:108)\n\tat software.amazon.awssdk.protocols.xml.internal.unmarshall.AwsXmlPredicatedResponseHandler.handle(AwsXmlPredicatedResponseHandler.java:85)\n\tat software.amazon.awssdk.protocols.xml.internal.unmarshall.AwsXmlPredicatedResponseHandler.handle(AwsXmlPredicatedResponseHandler.java:43)\n\tat software.amazon.awssdk.core.internal.handler.BaseClientHandler.lambda$successTransformationResponseHandler$7(BaseClientHandler.java:279)\n\tat software.amazon.awssdk.core.internal.http.async.AsyncResponseHandler.lambda$prepare$0(AsyncResponseHandler.java:92)\n\tat java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.complete(Unknown Source)\n\tat software.amazon.awssdk.core.internal.http.async.AsyncResponseHandler$BaosSubscriber.onComplete(AsyncResponseHandler.java:135)\n\tat software.amazon.awssdk.http.nio.netty.internal.ResponseHandler$DataCountingPublisher$1.onComplete(ResponseHandler.java:519)\n\tat software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.runAndLogError(ResponseHandler.java:254)\n\tat software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.access$600(ResponseHandler.java:77)\n\tat software.amazon.awssdk.http.nio.netty.internal.ResponseHandler$PublisherAdapter$1.onComplete(ResponseHandler.java:375)\n\tat software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerPublisher.publishMessage(HandlerPublisher.java:402)\n\tat software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerPublisher.flushBuffer(HandlerPublisher.java:338)\n\tat software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerPublisher.receivedDemand(HandlerPublisher.java:291)\n\tat software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerPublisher.access$200(HandlerPublisher.java:61)\n\tat software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerPublisher$ChannelSubscription$1.run(HandlerPublisher.java:495)\n\tat io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)\n\tat io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)\n\tat io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)\n\tat io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:566)\n\tat io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)\n\tat io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)\n\t... 1 common frames omitted\n"}

Seems something is wrong with S3: `Caused by: software.amazon.awssdk.services.s3.model.S3Exception: null (Service: S3, Status Code: 400, Request ID: null)` (classic null description from s3 client...)

Not sure what is wrong but we need to investigate.

Notes:
- the previous release with following commit patches https://github.com/linagora/tmail-backend/commit/8084bd34932647099b3cfd80fc7ade1e69c5768b and https://github.com/linagora/tmail-backend/commit/f5389f1c7dc10b2b03d097dbd2e77c9e9d1d20d0 was working for this jmap method
- the only other calendar commit related I see different from the patched previous release is this one: https://github.com/linagora/tmail-backend/commit/1dff2240d519d0987d5d905d8fb50370d6c81867 ... potential suspect?

--- a/tmail-backend/guice/distributed/src/main/java/com/linagora/tmail/event/RabbitMQAndRedisEventBusModule.java
+++ b/tmail-backend/guice/distributed/src/main/java/com/linagora/tmail/event/RabbitMQAndRedisEventBusModule.java
@@ -159,7 +159,7 @@ public class RabbitMQAndRedisEventBusModule extends AbstractModule {
             Configuration config = propertiesProvider.getConfiguration("redis");
             return RedisEventBusConfiguration.from(config);
         } catch (FileNotFoundException e) {
-            LOGGER.error("Missing `redis.properties` configuration file -> using default RedisEventBusConfiguration");
+            LOGGER.info("Missing `redis.properties` configuration file -> using default RedisEventBusConfiguration");
             return RedisEventBusConfiguration.DEFAULT;
         }
     }

--- a/tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/CalendarEventReplyPerformer.scala
+++ b/tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/CalendarEventReplyPerformer.scala
@@ -169,8 +169,9 @@ class CalendarEventMailReplyGenerator(val bodyPartContentGenerator: CalendarRepl
             .name(generateMailName())
             .sender(attendeeReply.attendee.asString())
             .addRecipients(recipient)
+            .mimeMessage(mimeMessage)
             .build()
-          mailImpl.setMessageNoCopy(mimeMessage)
+          LifecycleUtil.dispose(mimeMessage)
           mailImpl
         }))
       .subscribeOn(Schedulers.boundedElastic())
@@ -197,7 +198,7 @@ class CalendarEventMailReplyGenerator(val bodyPartContentGenerator: CalendarRepl
 
   private def generateMailName(): String = "calendar-reply-" + UUID.randomUUID().toString
 
-  private def appendAttachmentPartToMail(attachmentPart: Seq[BodyPartBuilder], mimeMessage: MimeMessageWrapper): MimeMessageWrapper = {
+  private def appendAttachmentPartToMail(attachmentPart: Seq[BodyPartBuilder], mimeMessage: MimeMessage): MimeMessage = {
     mimeMessage.getContent match {
       case mimeMultipart: MimeMultipart =>
         attachmentPart.foreach(part => mimeMultipart.addBodyPart(part.build))
@@ -216,7 +217,7 @@ class CalendarEventMailReplyGenerator(val bodyPartContentGenerator: CalendarRepl
     mimeMessage
   }
 
-  private def addHeaderToMimeMessage(toAddress: MailAddress, fromAddress: MailAddress, mimeMessage: MimeMessageWrapper): MimeMessageWrapper = {
+  private def addHeaderToMimeMessage(toAddress: MailAddress, fromAddress: MailAddress, mimeMessage: MimeMessage): MimeMessage = {
     mimeMessage.addFrom(Array(new InternetAddress(fromAddress.asString())))
     mimeMessage.addRecipients(Message.RecipientType.TO, toAddress.asString())
     mimeMessage.saveChanges()
@@ -225,7 +226,7 @@ class CalendarEventMailReplyGenerator(val bodyPartContentGenerator: CalendarRepl
 }
 
 trait CalendarReplyMessageGenerator {
-  def getBasedMimeMessage(i18n: Locale, attendeeReply: AttendeeReply, calendar: Calendar): SMono[MimeMessageWrapper]
+  def getBasedMimeMessage(i18n: Locale, attendeeReply: AttendeeReply, calendar: Calendar): SMono[MimeMessage]
 }
 
 private object I18NCalendarEventReplyMessageGenerator {
@@ -248,17 +249,17 @@ class I18NCalendarEventReplyMessageGenerator(fileSystem: FileSystem, i18nEmlDire
 
   import I18NCalendarEventReplyMessageGenerator._
 
-  override def getBasedMimeMessage(i18n: Locale, attendeeReply: AttendeeReply, requestCalendar: Calendar): SMono[MimeMessageWrapper] =
+  override def getBasedMimeMessage(i18n: Locale, attendeeReply: AttendeeReply, requestCalendar: Calendar): SMono[MimeMessage] =
     getBasedMimeMessage(attendeeReply, requestCalendar, evaluateMailTemplateFileName(attendeeReply.partStat, i18n))
 
-  def getBasedMimeMessage(attendeeReply: AttendeeReply, calendar: Calendar, emlFilename: String): SMono[MimeMessageWrapper] =
+  def getBasedMimeMessage(attendeeReply: AttendeeReply, calendar: Calendar, emlFilename: String): SMono[MimeMessage] =
     SMono.fromCallable(() => URI.create(i18nEmlDirectory).resolve(emlFilename).toString)
       .doOnNext(validateTemplateFile)
       .map(emlLocation => new MimeMessageWrapper(MimeMessageInputStreamSource.create(MailImpl.getId, fileSystem.getResource(emlLocation))))
       .map(mimeMessage => decoratedMimeMessage(mimeMessage, getMustacheDataMap(attendeeReply, calendar)))
       .subscribeOn(Schedulers.boundedElastic())
 
-  private def decoratedMimeMessage(originalMimeMessage: MimeMessageWrapper, mustacheDataMap: JavaMap[String, String]): MimeMessageWrapper = {
+  private def decoratedMimeMessage(originalMimeMessage: MimeMessageWrapper, mustacheDataMap: JavaMap[String, String]): MimeMessage = {
     originalMimeMessage.loadMessage()
     originalMimeMessage.getContent match {
       case textBody: String =>

--- a/tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/CalendarEventReplyPerformer.scala
+++ b/tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/CalendarEventReplyPerformer.scala
@@ -15,7 +15,7 @@ import com.linagora.tmail.james.jmap.model.{AttendeeReply, CalendarAttendeeField
 import eu.timepit.refined.auto._
 import jakarta.annotation.PreDestroy
 import jakarta.inject.{Inject, Named}
-import jakarta.mail.internet.{InternetAddress, MimeMultipart}
+import jakarta.mail.internet.{InternetAddress, MimeMessage, MimeMultipart}
 import jakarta.mail.{Message, Part}
 import net.fortuna.ical4j.model.Calendar
 import net.fortuna.ical4j.model.component.VEvent

